### PR TITLE
Handle symbol = symbol explicitly in stats

### PR DESCRIPTION
**Pull request by sopel39**

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/cost/TestFilterStatsCalculator.java [hunk 1]
```diff
@@ -485,6 +486,22 @@ public void testBetweenOperatorFilter()
     }
 
     @Test
+    public void testSymbolEqualsSymbolFilter()
```
Comment (losipiuk) [inline presto-main/src/test/java/com/facebook/presto/cost/TestFilterStatsCalculator.java]: rename to `testSymbolEqualsSameSymbolFilter()`

**Review by losipiuk — CHANGES_REQUESTED** [review]

File: presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java [hunk 2]
```diff
@@ -311,6 +311,14 @@ else if (!leftStats.isPresent() && rightStats.isPresent()) {
             if (leftStats.equals(UNKNOWN_STATS) || rightStats.equals(UNKNOWN_STATS)) {
                 return Optional.empty();
             }
+
+            if (left instanceof SymbolReference && right instanceof SymbolReference && left.equals(right)) {
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java:315]: When such a condition can be created?

File: presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java [hunk 3]
```diff
@@ -311,6 +311,14 @@ else if (!leftStats.isPresent() && rightStats.isPresent()) {
...
+                // any non-null value pass filter condition
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java]: "passes"

anyway, you'd better delegate to handling of `IS NOT NULL`

File: presto-main/src/test/java/com/facebook/presto/cost/TestFilterStatsCalculator.java [hunk 4]
```diff
@@ -215,8 +216,8 @@ public void testOrStats()
                                 .distinctValuesCount(20.0)
                                 .nullsFraction(0.0));
 
-        Expression leftExpressionSingleValue = new ComparisonExpression(ComparisonExpressionType.EQUAL, new SymbolReference("x"), new DoubleLiteral("0.0"));
-        Expression rightExpressionSingleValue = new ComparisonExpression(ComparisonExpressionType.EQUAL, new SymbolReference("x"), new DoubleLiteral("-7.5"));
+        Expression leftExpressionSingleValue = new ComparisonExpression(EQUAL, new SymbolReference("x"), new DoubleLiteral("0.0"));
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/cost/TestFilterStatsCalculator.java]: unrelated change (here & below). Each such a change should be a fixup to the right commit

File: presto-main/src/test/java/com/facebook/presto/cost/TestFilterStatsCalculator.java [hunk 5]
```diff
@@ -266,10 +267,10 @@ public void testAndStats()
                                 .distinctValuesCount(15.0)
                                 .nullsFraction(0.0));
 
-       // Impossible, with symbol-to-expression comparisons
+        // Impossible, with symbol-to-expression comparisons
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/cost/TestFilterStatsCalculator.java]: should be a fixup

File: presto-main/src/test/java/com/facebook/presto/cost/TestFilterStatsCalculator.java [hunk 6]
```diff
@@ -485,6 +486,22 @@ public void testBetweenOperatorFilter()
     }
 
     @Test
+    public void testSymbolEqualsSymbolFilter()
+    {
+        Expression symbolEqualsSymbol = new ComparisonExpression(EQUAL, new SymbolReference("x"), new SymbolReference("x"));
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/cost/TestFilterStatsCalculator.java]: I think we shouldn't have unit test for that, as the (more) correct solution would be to fix the party responsible for introducing/retaining such (almost) useless comparison. We _do_ need an integration test, however, e.g. in `TestTpchLocalStats`

**Review by findepi — CHANGES_REQUESTED** [review]

File: presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java (hunk 2 above)
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java:315]: you know by now...

Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java:315]: I still don't :) I tried, but i failed.

File: presto-main/src/test/java/com/facebook/presto/cost/TestFilterStatsCalculator.java (hunk 6 above)
Comment (sopel39) [reply presto-main/src/test/java/com/facebook/presto/cost/TestFilterStatsCalculator.java]: I wouldn't call `a=a` is a bug. It is a perfectly legal expression. If we can test it on unit level (even as a special case), then why not?

**Comment by sopel39** [comment]
applied comments

**Review by losipiuk — APPROVED** [review]

**Pull request closed without merging**